### DISCUSSION TITLE
[Uno.Wasm.Bootstrap] Call C# methods from JS and vice-verca

### DISCUSSION THREAD

**Discussion by Vawlpe** [unoplatform/uno#8548] [Q&A] [answered]
Just a quick question, I only need UNO for the `Uno.Wasm.Bootstrap`, I don't want the UI framework or anything else, so I set up a simple project according to the instructions in the `README.md` on `Uno.Wasm.Bootstrap`, so now, out of curiosity how would I actually call a C# method in my bootstrapped project from JS, as well as the opposite, call JS functions from within my C# code?

I'd theoretically want to pass a `WebGLRenderingContext` from JS to C#, then write a wrapper in C# that takes the WebGL context API methods and call their JS equivalents, basically allowing me to build a WebGL program in C#

**Comment by jeromelaban** [comment, marked as answer]
If you're using the bootstrapper directly, the only way to handle this is through [Javascript evals](https://github.com/unoplatform/Uno.Wasm.Bootstrap/blob/bd539464da2b9f6a72ec3a64542b557ac0725cf1/src/Uno.Wasm.AotProfiler/Runtime.cs), that is if you want to use the Javascript API.

You could also try to use the native WebGL API, using [P/Invokes](https://github.com/unoplatform/Uno.Wasm.Bootstrap#webassembly-module-linking-support) and specifically the ability to [invoke emscripten symbols](https://github.com/unoplatform/Uno.Wasm.Bootstrap#static-linking-additional-pinvoke-libraries).

Calling back is a bit more complex, but you [can do it this way](https://github.com/unoplatform/uno/blob/124143558f3ccba97a17dfe90a2fe576bb581d77/src/Uno.UI/ts/Windows/Storage/StorageFolder.ts#L80).

**Reply by Vawlpe** [reply]
Thx a lot for the info~! I tried to figure some of that out but I have no experience with WASM, or Uno, or anything in the same ballpark, so this is all very confusing honestly, I couldn't figure out how the js eval stuff you linked would work anywhere outside of the environment it was presented in, I also tried the P/Invoke stuff, trying to just call a `libc` function to see if I could get it working, but the moment I added the `WasmShellNativeCompile` property to the csproj and tried to build (regardless of whether or not I tried P/Invoke-ing anything) it'd error out asking for WSL (which I do have, but it thought it didn't for some reason), as for the callback example... I didn't even try, it just seemed pointless if I couldn't get the other 2 to work whatsoever, and I had the same problem of not knowing how to recontextualize the example xD

I've also tried other alternatives to get my code compiled to WASM and interacting with WebGL, other than the Uno bootstrap, initially, I just wanted to compile my code to wasm with [NativeAOT-LLVM](https://github.com/dotnet/runtimelab/tree/feature/NativeAOT-LLVM) but I couldn't exactly figure out how to pass variables back and forth between JS and my WASM compiled C# code with that, I also tried the now archived WebGL.Net package from Wave Engine, but that didn't exactly go well either, and at this point, I'm slowly considering porting my entire codebase to C++ or giving up on ever getting my project to run in a browser :/

**Reply by jeromelaban** [reply]
There's no direct integration with WebGL at this point, but maybe there's a way to get Silk.NET to resolve the imports properly. You're in uncharted territory at this point, so it not at straight path yet. There was something the WaveEngine team [did a while back](https://www.plainconcepts.com/wave-engine-mono-wasm/).

**Reply by Vawlpe** [reply]
Sidenote, I also found [this example](https://github.com/zwcloud/MinimalCSharpWasmWebGL-NativeAOT-LLVM) of WebGL running in WASM compiled with NativeAOT-LLVM, however it seems to be executing all the WebGL code in c++ and just P/Invoking an initialization function it defines... now that I think about it, that wouldn't be much of a problem, I can just do the same thing myself... why did I waste so much time trying all this other stuff lmfao, lemme see if i can get that example working for my usecase ig lol

**Reply by jeromelaban** [reply]
True. This sample is not really a C# piece of code :)

**Reply by Vawlpe** [reply]
> There's no direct integration with WebGL at this point, but maybe there's a way to get Silk.NET to resolve the imports properly. You're in uncharted territory at this point, so it not at straight path yet. There was something the WaveEngine team [did a while back](https://www.plainconcepts.com/wave-engine-mono-wasm/).

Apparently, Silk.NET is working on WebGL and WebGPU targets for 3.0, which are supposed to work on top of Blazor Wasm, so that's cool ig :p
